### PR TITLE
[#7006] login with keycard

### LINKED ISSUE #7006: Login with keycard
# Problem

It's not possible to login into Status with keycard.

# Implementation

TBD. Related work needs to be done on status-go side https://github.com/status-im/status-go/compare/feature/keycard-account

# Acceptance Criteria

It's should be possible to login with keycard into the account created with keycard.

# Notes

Expirementing already started here https://github.com/status-im/status-react/tree/feature/login-with-keycard

# Future Steps

Use status-go `LoginWithKeycard` once API will be finalized https://github.com/status-im/status-go/compare/feature/keycard-account

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-go">status-im/status-go</a></strong></div><div>The Status module that consumes go-ethereum. Contribute to status-im/status-go development by creating an account on GitHub.</div></blockquote>

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #7006 

### Summary:

Implements login with keycard

#### Platforms
- Android

status: ready

### CODE REVIEW

File: STATUS_GO_VERSION [hunk 1]
```diff
@@ -1 +1 @@
-0.19.0-beta.9
+190124-170357-929a5d
```
Comment (dmitryn) [inline STATUS_GO_VERSION]: It will be changed after new status-go library version will be released

File: src/status_im/accounts/login/core.cljs [hunk 2]
```diff
@@ -40,8 +40,13 @@
 
 ;;;; Handlers
 (fx/defn login [cofx]
-  (let [{:keys [address password]} (accounts.db/credentials cofx)]
-    {:accounts.login/login [address password]}))
+  (if (get-in cofx [:db :hardwallet :whisper-private-key])
```
Comment (flexsurfer) [inline src/status_im/accounts/login/core.cljs:44]: `:whisper-private-key` sounds scary?

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/accounts/login/core.cljs (hunk 2 above)
Comment (dmitryn) [reply src/status_im/accounts/login/core.cljs:44]: @flexsurfer can you suggest better name? It's actually a private key for whisper

**Comment by denis-sharypin** [comment]
@guylouis I found out that there is already a PR with the behavior that was outlined in Figma file (Tap with a card then enter a PIN). How we should proceed? It will take time to create a new logic and add fixes to UI.

File: src/status_im/accounts/login/core.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/status_im/accounts/login/core.cljs:44]: it's not about name, just wonder if it's save to store private key in app-db

**Comment by guylouis** [comment]
@dmitryn the point is about when do we ask the PIN to the user for the login, and how many taps it takes to login. 

 I see it as we have two steps for login
1/ first the account must be identified
      1-a/ by selecting it manually among a list
      1-b/ by tapping the card on the phone

2/ PIN validation

Can we do the following implementation ?
1-a + 2 : only one tap overall 
1-b + 2 : two taps 

The spec as defined des not go in this level of detail and considers we're always 1-a + 2 
https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows?node-id=0%3A1




<blockquote><div> Figma</div><div><strong><a href="https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows">Figma</a></strong></div><div>Created with Figma</div></blockquote>

File: src/status_im/accounts/login/core.cljs (hunk 2 above)
Comment (dmitryn) [reply src/status_im/accounts/login/core.cljs:44]: @flexsurfer we will remove it from app-db after we pass it to status-go.

**Comment by dmitryn** [comment]
@guylouis yes, we do 1-a + 2.

User selects account from the list, then we ask him to tap the card and then ask for PIN.

One more scenario with faster flow: If card is already close enough to the phone (i.e. we have connection), we don't ask to tap, we directly go to the PIN screen after user selected account from the list.

**Comment by guylouis** [comment]
you mean we will do 1-b later ?

**Comment by dmitryn** [comment]
@guylouis no, 1-b is also supported. On accounts page, when user taps the card, we found corresponding account based on keycard instance uid and then process to login.

**Comment by guylouis** [comment]
ok perfect. 
In the case 1-a + 2, can't we cache the PIN (just like transaction signing) to avoid the user to tap two times the card but only one time ?

**Comment by dmitryn** [comment]
@guylouis we can cache the PIN, but user inputs PIN only once during the login.

**Comment by guylouis** [comment]
@Dmitry Novotochinov <dmitry@status.im> yes, I mean caching the PIN to
avoid to have 2 taps.

**Review by goranjovic — APPROVED** [review]

**Comment by dmitryn** [comment]
@asemiankevich android build is ready http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190130-120842-06abc6-pr.apk

<blockquote></blockquote>

**Comment by antdanchenko** [comment]
@dmitryn  please rebase the PR on develop in order to include fix from #7381

**Comment by dmitryn** [comment]
@antdanchenko ok, rebased and pushed

**Comment by asemiankevich** [comment]
@dmitryn 

1. [Plus] button is always spinning when i create account with Keycard and logged in right after account creation:

image

2. Card is never detected when i try to login with a keycard. 

image


```
2019-01-31 11:41:08.336 1154-1154/? I/TrafficMonitor: update:rxPkts:0,txPkts:0,rxBytes:0,txBytes:0
2019-01-31 11:41:08.336 1154-1154/? I/TrafficMonitor: start expired. level:-1
2019-01-31 11:41:08.336 1154-1154/? I/TrafficMonitor: gettimer:interval=2000
2019-01-31 11:41:08.352 670-3847/? D/NxpTml: PN54X - I2C Read successful.....len = 31
2019-01-31 11:41:08.353 670-3847/? D/NxpTml: PN54X - Posting read message.....
2019-01-31 11:41:08.353 670-3850/? D/NxpHal: read successful status = 0x0
2019-01-31 11:41:08.353 670-3850/? D/NxpHal: NxpNci: RF Interface = ISO-DEP
2019-01-31 11:41:08.353 670-3850/? D/NxpHal: NxpNci: Protocol = ISO-DEP
2019-01-31 11:41:08.353 670-3850/? D/NxpHal: NxpNci: Mode = A Passive Poll
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: NFC received ntf gid:1
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: returning from nfa_dm_p2p_prio_logic  Disable p2p_prio_logic
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfc_set_state 4 (IDLE)->5 (OPEN)
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfc_ncif_proc_activate:28 5, mode:0x00
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_disc_discovery_cback (): event:0x4004
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_disc_sm_execute (): state: DISCOVERY (1), event: ACTIVATED_NTF(5) disc_flags: 0x1
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_disc_new_state (): old_state: DISCOVERY (1), new_state: POLL_ACTIVE (4) disc_flags: 0x1
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_disc_notify_activation (): tech_n_mode:0x0, proto:0x4
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_disc_get_disc_mask (): tech_n_mode:0x0, protocol:0x4, disc_mask:0x4
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: activated_protocol:0x4, activated_handle: 0x1
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_poll_disc_cback (): event:0x01
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_rw_handle_event event: NFA_RW_ACTIVATE_NTF_EVT (0x601), flags: 00000001
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_rw_activate_ntf
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: RW_SetActivatedTagType protocol:4, technology:0, SAK:32
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: rw_t4t_select ()
2019-01-31 11:41:08.353 3626-3821/? I/BrcmNfcNfa: nfa_dm_notify_activation_status (): status:0x0
2019-01-31 11:41:08.353 3626-3821/? D/BrcmNfcJni: nfaConnectionCallback: Connection Event = 5
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: nfaConnectionCallback: NFA_ACTIVATED_EVT: gIsSelectingRfInterface=0, sIsDisabling=0
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: checkforTranscation: enter; event=0x5 transaction_data.current_transcation_state = 0x0
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: checkforTranscation: exit; event=0x5 transaction_data.current_transcation_state = 0x0
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::setActivationState: state=2
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: pn544InteropIsBusy: 0
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: nativeNfcTag_handleNonNciMultiCardDetection: enter 
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::IsSameKovio: enter
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::discoverTechnologies (activation): enter
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: mNumTechList =0, mTechListIndex=0
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::discoverTechnologies (activation): index=0; tech=3; handle=1; nfc type=4
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::discoverTechnologies (activation): index=1; tech=1; handle=1; nfc type=4
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::discoverTechnologies (activation): exit
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::createNativeNfcTag: enter
2019-01-31 11:41:08.354 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers1
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers2
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers3: index=0; rf tech params mode=0
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers3: tech A
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers3: index=1; rf tech params mode=0
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers3: tech A
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers4: index=0
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers4: T4T; ISO_DEP for tech A; copy historical bytes; len=0
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers4: index=1
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers4: T4T; tech A
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers5: tech A
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::fillNativeNfcTagMembers5;mTechListIndex=0
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::createNativeNfcTag; mNumDiscNtf=0
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::createNativeNfcTag: try notify nfc service
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: deleteglobaldata: enter
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: deleteglobaldata: exit
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: NfcTag::createNativeNfcTag: exit
2019-01-31 11:41:08.355 670-3847/? D/NxpTml: PN54X - Read requested.....
2019-01-31 11:41:08.355 670-3850/? D/NxpHal: phNxpNciHal_client_thread: CHR_No-7
2019-01-31 11:41:08.355 670-3847/? D/NxpTml: PN54X - Invoking I2C Read.....
2019-01-31 11:41:08.355 3626-3821/? D/BrcmNfcJni: nfaConnectionCallback: scoreGenericNtf = 0x0
2019-01-31 11:41:08.355 3626-3821/? I/BrcmNfcNfa: nfa_dm_disc_sm_execute (): new state: POLL_ACTIVE (4), disc_flags: 0x1
2019-01-31 11:41:08.355 3626-3822/? I/BrcmNfcNfa: nfc_main_hal_cback event: HAL_NFC_CHR_EVT(0xf0), status=7
2019-01-31 11:41:08.355 3626-3822/? I/BrcmNfcNfa: nfa_dm_nfc_response_cback () NFC_CHR_REVT(0x5015)
2019-01-31 11:41:08.355 3626-3822/? D/BrcmNfcJni: nfaDeviceManagementCallback: enter; event=0xD
2019-01-31 11:41:08.355 3626-3822/? E/BrcmNfcJni: nfaDeviceManagementCallback: chr_sub_error_no:7
2019-01-31 11:41:08.356 3626-3735/? E/NfcNci: HwNfcDFTConnManager error: Not in Filed or not commu data
2019-01-31 11:41:08.357 1154-1680/? W/ActivityManager: Sending non-protected broadcast com.huawei.nfc.action.NDEF_DISCOVERED from system 3626:com.android.nfc/1027 pkg com.android.nfc
2019-01-31 11:41:08.359 3626-3626/? I/SoundPool: Pid:3626 SoundChannel play
```

3. Do we want to skip the [Enter username] screen when creating account with keycard?

image

4. Can't logout from an account which was created with a password. Steps:
- open status -> password -> create account with password
- profile -> tap logout

5. can't login offline. Steps:
- turn on airplane mode
- open status

As result: blank screen appears. I can't launch the app so forced to reinstall it.

**Comment by dmitryn** [comment]
@asemiankevich thanks for testing, i'm looking into each issue you found.

**Review by jakubgs — APPROVED** [review]
`mobile_files/yarn.lock` changes look fine to me.

**Comment by dmitryn** [comment]
@asemiankevich issues are fixed, please try new build http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190201-124557-dc6170-pr.apk

Also, login flows in Figma have changed https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows?node-id=0%3A1


<blockquote></blockquote>
<blockquote><div> Figma</div><div><strong><a href="https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows">Figma</a></strong></div><div>Created with Figma</div></blockquote>

**Comment by asemiankevich** [comment]
> @asemiankevich issues are fixed, please try new build http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190201-124557-dc6170-pr.apk
> 
> Also, login flows in Figma have changed https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows?node-id=0%3A1
> 
> >  Figma**[Figma](https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows)**Created with Figma

i believe we still should have screens when enter user name right after creation account with keycard, right? @dmitryn 

<blockquote></blockquote>
<blockquote><div> Figma</div><div><strong><a href="https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Keycard-flows">Figma</a></strong></div><div>Created with Figma</div></blockquote>

**Comment by dmitryn** [comment]
@asemiankevich no, we don't have those screens in Figma so we don't ask user name after account creation.

**Comment by guylouis** [comment]
@denis-sharypin 
creating a name should be the same if the user has or not a Keycard right ?

**Comment by asemiankevich** [comment]
@dmitryn when i try to login with keycard account PIN validation fails. Also NFC connection is lost each attempt i enter PIN and then i see `invalid PIN` error.
@antdanchenko why e2e are failed?

**Comment by antdanchenko** [comment]
@asemiankevich the best way to find why they failed is to watch video record of those tests, e.g.
[here you can see blank screen after opening the app in offline mode](https://app.saucelabs.com/tests/bf2a5c68b95340b797b524301e006f56/watch?auth=256d0491ffaa1d9c18f4fd520344769b#54)
or [disabled logout button instead of login screen after logging out ](https://app.saucelabs.com/tests/05f78f1fbc78405caf34b635f56ae385/commands?auth=6694354d40b63a836da39866323208b3#54) on another test
video can be found under each test, just check `Steps, video, logs` links

**Comment by dmitryn** [comment]
@asemiankevich i've made fixes, please try new build http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190204-101734-7dc019-pr.apk

<blockquote></blockquote>

**Comment by asemiankevich** [comment]
@dmitryn:

There are 2 problems in the PR found by autotests:
- blank screen if i close the app and open it again. Steps: open app - create account - close the app - open the app
- can't logout. Logout button is disabled and nothing happens.

Could you take a look? I checked in nightly and it is fine there

**Comment by dmitryn** [comment]
@asemiankevich sure, i'll take a look

**Comment by ghost** [comment]
`mobile_files/yarn.lock` changed. Pinging @jakubgs, @mandrigin, and @corpetty

**Comment by dmitryn** [comment]
@asemiankevich i've fixed the issues from e2e tests, here is new build: http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190205-100356-225b15-pr.apk

Not sure why e2e tests were not run for that build though. 

<blockquote></blockquote>

**Comment by asemiankevich** [comment]
looks good to me! 2 minor points i would like to address:

1. no `thanks for choosing beta..` pop -up when creating account with keycard
2. spinner takes ~ 4 seconds before user is logged in to the keycard account 
image

@denis-sharypin do we want to make is a bit more visible? like something `logging in..` in the middle of the screen? First time i even did not notice the spinner :D 

All those could be (if needed) addressed in separate issues. Lets merge this PR

**Comment by corpetty** [comment]
Looks good guys, great work.

**Pull request merged**